### DISCUSSION TITLE
installation error

### DISCUSSION THREAD

**Discussion by srbrodzik** [ARM-DOE/pyart#1651] [General]
I'm trying to upgrade my pyart installation (on a Debian system running version 10) by downloading the zip file, unzipping and running

python setup.py install

When I do so, I get a syntax error that I don't know what to do with

  File "setup.py", line 79
    *sys.version_info[:2], *min_version
                           ^
SyntaxError: invalid syntax

Suggestions?  I've attached the setup.py file with a txt suffix since I can't attach a py file.

[setup.txt](https://github.com/user-attachments/files/17049542/setup.txt)

**Comment by zssherman** [comment]
@srbrodzik What Python version is on that system?

**Comment by srbrodzik** [comment]
python3.7

**Reply by mgrover1** [reply]
The latest version(s) of Py-ART require at least Python 3.9, can you please update your Python version and retry?

**Comment by srbrodzik** [comment]
I'll do that and let you know how it goes.

**Reply by mgrover1** [reply]
Thanks! Thank you for opening this discussion!

**Reply by srbrodzik** [reply]
So I upgraded to Debian 11 and python 3.9 and tried installing the newest pyart again.  Still having problems running this command from the pyart-main directory:

python3 setup.py install

Traceback (most recent call last):
  File "/home/disk/meso-home/brodzik/python/pyart-main/setup.py", line 297, in <module>
    ext_modules=cythonize(
  File "/usr/lib/python3/dist-packages/Cython/Build/Dependencies.py", line 961, in cythonize
    c_options = CompilationOptions(**options)
  File "/usr/lib/python3/dist-packages/Cython/Compiler/Main.py", line 573, in __init__
    raise ValueError(message)
ValueError: got unknown compiler directive: cpow

**Comment by srbrodzik** [comment]
Did you see my latest comment?  The install still isn't working.

**Reply by zssherman** [reply]
@srbrodzik Apologies, I missed that last comment. Could you try updating your cython version? Might be an issue there.

**Comment by srbrodzik** [comment]
I updated cython to v3.1.11.  That seemed to work and the setup.py install worked.  I did it locally (to my ~/.local/lib dir.  But when I tried to import it, I got several messages including one about numpy.  I'm currently using v2.0.2.

shear:brodzik:119>ipython3
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.20.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pyart

## You are using the Python ARM Radar Toolkit (Py-ART), an open source
## library for working with weather radar data. Py-ART is partly
## supported by the U.S. Department of Energy as part of the Atmospheric
## Radiation Measurement (ARM) Climate Research Facility, an Office of
## Science user facility.
##
## If you use this software to prepare a publication, please cite:
##
##     JJ Helmus and SM Collis, JORS 2016, doi: 10.5334/jors.119

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-413608f118be> in <module>
----> 1 import pyart

~/.local/lib/python3.9/site-packages/arm_pyart-1.19.1.post8-py3.9-linux-x86_64.egg/pyart/__init__.py in <module>
     27 # import subpackages
     28 # print out helpful message if build fails or importing from source tree
---> 29 from . import (
     30     __check_build,  # noqa                                                                                  
     31     aux_io,  # noqa

~/.local/lib/python3.9/site-packages/arm_pyart-1.19.1.post8-py3.9-linux-x86_64.egg/pyart/aux_io/__init__.py in <module>                                                                                                                 
     14 """
     15 
---> 16 from .arm_vpt import read_kazr  # noqa
     17 from .d3r_gcpex_nc import read_d3r_gcpex_nc  # noqa
     18 from .edge_netcdf import read_edge_netcdf  # noqa

~/.local/lib/python3.9/site-packages/arm_pyart-1.19.1.post8-py3.9-linux-x86_64.egg/pyart/aux_io/arm_vpt.py in <module>                                                                                                                  
      7 """
      8 
----> 9 import netCDF4
     10 import numpy as np
     11 

/usr/local/lib/python3.9/dist-packages/netCDF4/__init__.py in <module>
      1 # init for netCDF4. package
      2 # Docstring comes from extension module _netCDF4.
----> 3 from ._netCDF4 import *
      4 # Need explicit imports for names beginning with underscores
      5 from ._netCDF4 import __doc__

src/netCDF4/_netCDF4.pyx in init netCDF4._netCDF4()

/usr/lib/python3/dist-packages/cftime/__init__.py in <module>
----> 1 from ._cftime import datetime, real_datetime, _parse_date
      2 from ._cftime import num2date, date2num, date2index, time2index, num2pydate
      3 from ._cftime import microsec_units, millisec_units, \
      4                      sec_units, hr_units, day_units, min_units,\
      5                      UNIT_CONVERSION_FACTORS

src/cftime/_cftime.pyx in init cftime._cftime()

ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject

**Reply by kmuehlbauer** [reply]
@srbrodzik That looks like some real mess with the environments.

It seems that cftime is pulled in from /use/lib/python3 whereas netCDF4 is taken from /use/local/lib/python3.9

That error you are seeing indicates such environment clash.

**Comment by srbrodzik** [comment]
Thinking numpy was the culprit, I went back to v1.19.5.  Then I got a message that I needed at least numpy1.22.4 to be compatible with pandas (v2.2.3).  Now getting a new numpy error:

ModuleNotFoundError: No module named 'numpy._typing'

Please let me know what version of numpy I need to have installed.

**Comment by zssherman** [comment]
@srbrodzik I agree with @kmuehlbauer that I think this is an environment issue with a few conflicts happening. The newest numpy would be better as it contains changes that Py-ART has in it, so older versions might not work. With the newer numpy you could also try updating your netcdf and cftime or cleaning the environment and starting fresh if that's possible? I'm trying to think of other options. Using an environment maintainer like mamba might be an option as well.

**Reply by srbrodzik** [reply]
We decided to do our own make and installation.  It finished but then complained that we needed xradar.  Getting that to install is proving to be a major pain in the butt.  Any hints would be appreciated.

**Reply by zssherman** [reply]
@srbrodzik What kind of issues are you running into for xradar?

**Reply by srbrodzik** [reply]
They switched entirely to one of the new install/builds for python. This would be fine if everyone did it, and everyone was forced to a version of pip that deals with it. However, most packaging systems are not their yet, so continuing to support the old setup.py method while this is occurring would make sense. It so happens that the debian python group has a test version of support for this, but others do not yet.

**Reply by kmuehlbauer** [reply]
Setuptools 61.0.0 was released on March 24th, 2022. It has full capabilities to handle the pyproject.toml based builds. 

Xradar was conceived in late August 2022 and used the pyproject.toml based builds right from the start.

Now, approaching end of 2024, it's hard to believe that there are still packaging systems which do not support standards that are out there for almost four years (PEP 621) and are available in standard python setuptools for 2.5 years now.
